### PR TITLE
Delay removal of orphanPVC to avoid the removal of PVC in use

### LINKED ISSUE #32: deleteOrphanPvc deleted PVC in use
I'm trying to create a Druid cluster using druid-operator on AWS EKS. I'm using EBS GP2 for the persistent volume. 

When trying to scale up the historical pods (e.g. 4 to 8), the first pod stuck in pending, and the rest 7 pods working fine. The first pvc was mistakenly deleted as orphan PVC even though it is still in use. 

**druid-operator log:**
1.6798315940261655e+09  INFO    druid_operator_handler  Deleted orphaned pvc [data-volume-druid-workload-historicals-4:default] successfully    {"name": "workload", "namespace": "default"}
1.679831594026486e+09   DEBUG   events  Normal  {"object": {"kind":"Druid","namespace":"default","name":"workload","uid":"2c6b92b9-73cb-408f-a670-a3ee7fc307ff","apiVersion":"druid.apache.org/v1alpha1","resourceVersion":"3088566"}, "reason": "DruidOperatorDeleteSuccess", "message": "Successfully deleted object [data-volume-druid-workload-historicals-4:PersistentVolumeClaim] in namespace [default]"}

This issue is reproducible in the following environments:
druid-operator (0.0.9), kubernetes (1.23). 

**Storage Class:**
Name:            gp2
IsDefaultClass:  Yes
Annotations:     kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"storage.k8s.io/v1","kind":"StorageClass","metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"},"name":"gp2"},"parameters":{"fsType":"ext4","type":"gp2"},"provisioner":"kubernetes.io/aws-ebs","volumeBindingMode":"WaitForFirstConsumer"}
,storageclass.kubernetes.io/is-default-class=true
Provisioner:           kubernetes.io/aws-ebs
Parameters:            fsType=ext4,type=gp2
AllowVolumeExpansion:  <unset>
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     WaitForFirstConsumer
Events:                <none>

### PR DESCRIPTION
**Pull request by gurjotkaur20**
Fixes #32 

### Description

Possibly a race condition is causing removal of PVC in use.

The solution in the PR: Added two labels `toBeDeleted` and `deletionTimestamp` to mark the pvc for deletion. Do not delete the pvc right away, wait for 60s to delete it and avoid the race condition.

<hr>

This PR has:
- [ ] been tested on a real K8S cluster to ensure creation of a brand new Druid cluster works.
- [ ] been tested for backward compatibility on a real K*S cluster by applying the changes introduced here on an existing Druid cluster. If there are any backward incompatible changes then they have been noted in the PR description.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added documentation for new or modified features or behaviors.

<hr>

##### Key changed/added files in this PR
 * `controllers/druid/handler.go`
 * `controllers/druid/util.go`

### CODE REVIEW

**Comment by AdheipSingh** [comment]
@gurjotkaur20 this is great !

The only thing which i might want to do is add a finalizer instead of labels ( cc @avtarOPS )  Regardless, ill take look at it.
Thanks

File: controllers/druid/handler.go [hunk 1]
```diff
@@ -490,12 +492,65 @@ func deleteOrphanPVC(ctx context.Context, sdk client.Client, drd *v1alpha1.Druid
 		for i, pvc := range pvcList {
 
 			if !ContainsString(mountedPVC, pvc.GetName()) {
-				err := writers.Delete(ctx, sdk, drd, pvcList[i], emitEvents, &client.DeleteAllOfOptions{})
-				if err != nil {
-					return err
+
+				if _, ok := pvc.GetLabels()["toBeDeleted"]; ok {
+					deletionTS := pvc.GetLabels()["deletionTimestamp"]
```
Comment (avtar-mindgrep) [inline controllers/druid/handler.go:497]: we can change the label name as the same label is used by the finalizer.

File: controllers/druid/handler.go [hunk 2]
```diff
@@ -490,12 +492,65 @@ func deleteOrphanPVC(ctx context.Context, sdk client.Client, drd *v1alpha1.Druid
 		for i, pvc := range pvcList {
 
 			if !ContainsString(mountedPVC, pvc.GetName()) {
-				err := writers.Delete(ctx, sdk, drd, pvcList[i], emitEvents, &client.DeleteAllOfOptions{})
-				if err != nil {
-					return err
+
```
Comment (itamar-marom) [inline controllers/druid/handler.go:495]: This became way too big, we should refactor this, separating it into functions

File: controllers/druid/handler.go [hunk 3]
```diff
@@ -490,12 +492,65 @@ func deleteOrphanPVC(ctx context.Context, sdk client.Client, drd *v1alpha1.Druid
...
+				if _, ok := pvc.GetLabels()["toBeDeleted"]; ok {
```
Comment (itamar-marom) [inline controllers/druid/handler.go:496]: This and the below should be consts

**Comment by gurjotkaur20** [comment]
I have refactored the code and updated the label name.

**Review by itamar-marom — APPROVED** [review]

**Comment by itamar-marom** [comment]
> I have refactored the code and updated the label name.

Fine by me :) 
@AdheipSingh

File: controllers/druid/handler.go [hunk 4]
```diff
@@ -490,19 +494,94 @@ func deleteOrphanPVC(ctx context.Context, sdk client.Client, drd *v1alpha1.Druid
...
+					getPvcLabels := pvc.GetLabels()
+					getPvcLabels[toBeDeletedLabel] = "yes"
+					getPvcLabels[deletionTSLabel] = strconv.FormatInt(time.Now().Unix(), 10)
+
+					err = setPVCLabels(ctx, sdk, drd, emitEvents, pvcList[i], getPvcLabels, true)
+					if err != nil {
+						return err
+					} else {
+						// do nothing
+					}
```
Comment (AdheipSingh) [inline controllers/druid/handler.go:514]: why do we need this block ?

File: controllers/druid/handler.go [hunk 5]
```diff
@@ -490,19 +494,94 @@ func deleteOrphanPVC(ctx context.Context, sdk client.Client, drd *v1alpha1.Druid
...
+					getPvcLabels := pvc.GetLabels()
+					delete(getPvcLabels, toBeDeletedLabel)
+					delete(getPvcLabels, deletionTSLabel)
+
+					err = setPVCLabels(ctx, sdk, drd, emitEvents, pvcList[i], getPvcLabels, false)
+					if err != nil {
+						return err
+					} else {
+						// do nothing
+					}
```
Comment (AdheipSingh) [inline controllers/druid/handler.go:528]: same, why do we need this block ? can return nil ?

**Review by AdheipSingh — APPROVED** [review]

File: controllers/druid/handler.go (hunk 5 above)
Comment (gurjotkaur20) [reply controllers/druid/handler.go]: > same, why do we need this block ? can return nil ?

Done changes for both else blocks.

**Review by himadrisingh — APPROVED** [review]

**Pull request merged**